ocs_ci/deployment/cnv.py

```diff
@@ -0,0 +1,510 @@
+            if (
+                os_type.lower() in text.lower()
+                and os_machine_type.lower() in text.lower()
+            ):
+                logger.info(
+                    f"The virtctl download URL for {os_type.lower()} {os_machine_type.lower()} is: {link_info['href']}"
+                )
+                return href
+        return None
+
+    def check_virtctl_compatibility(self):
+        """
+        Check if the virtctl binary is compatible with the current system.
+
+        Raises:
+            exceptions.ArchitectureNotSupported: If virtctl is not compatible.
+
+        """
+        logger.info("Performing virtctl binary compatibility check...")
+        os_type = platform.system()
+        os_machine_type = platform.machine()
+        virtctl_download_url = self.get_virtctl_download_url(
+            os_type=os_type, os_machine_type=os_machine_type
+        )
+        os_machine_type = "amd64" if os_machine_type == "x86_64" else os_machine_type
+        if (
+            not virtctl_download_url
+            or (os_type.lower() not in virtctl_download_url)
+            or (os_machine_type.lower() not in virtctl_download_url)
+        ):
+            raise exceptions.ArchitectureNotSupported(
+                f"Virtctl is NOT compatible to run on this machine: {os_type} {platform.machine()}"
+            )
+
+        logger.info(
+            f"Virtctl is compatible to run on this machine: {os_type} {platform.machine()}"
+        )
+
+    def download_and_extract_virtctl_binary(self, bin_dir=None):
+        """
+        Download and extract the virtctl binary to bin_dir
+
+        Args:
+            bin_dir (str): The directory to store the virtctl binary.
+
+        """
+        os_type = platform.system()
+        os_machine_type = platform.machine()
+        self.check_virtctl_compatibility()
+
+        # Prepare bin directory for virtctl
+        bin_dir_rel_path = os.path.expanduser(bin_dir or config.RUN["bin_dir"])
+        bin_dir = os.path.abspath(bin_dir_rel_path)
+        virtctl_binary_path = os.path.join(bin_dir, "virtctl")
+        if os.path.isfile(virtctl_binary_path):
+            logger.info(
+                f"virtctl binary already exists {virtctl_binary_path}, skipping download."
+            )
+        else:
+            (
+                archive_file_binary_data,
+                virtctl_download_url,
+            ) = self._download_virtctl_archive(os_type, os_machine_type)
+            self._extract_virtctl_binary(
+                archive_file_binary_data, virtctl_download_url, bin_dir
+            )
+
+    def _download_virtctl_archive(self, os_type, os_machine_type):
+        """
+        Download the virtctl binary archive.
+
+        Args:
+            os_type (str): The operating system.
+            os_machine_type (str): The operating system machine architecture.
+
+        Returns:
+        Tuple(io.BytesIO, str): Binary data of the downloaded archive and the virtctl download URL.
```

Comment:
@PrasadDesala , we need a tab here.
```suggestion
Returns:
    Tuple(io.BytesIO, str): Binary data of the downloaded archive and the virtctl download URL.
```